### PR TITLE
cdrs-tokio test suite integration

### PR DESCRIPTION
**Pull request by conorbros**

### CODE REVIEW

**Comment by conorbros** [comment]
Not ready for review yet, just trying to get the CI to run so I can check it's working.

**Comment by benbromhead** [comment]
This seems stuck, are we pursuing this (or a todo later)?

File: shotover-proxy/Cargo.toml
Comment (rukai) [inline shotover-proxy/Cargo.toml:14]: The `cpp` is a bit confusing, maybe one of:
* `cassandra-cpp-driver-tests` be specific about which driver we are enabling
* `external-dependency-integration-tests` be as generic as possible

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:303]: Is this related to this PR?

File: shotover-proxy/Cargo.toml
Comment (rukai) [inline shotover-proxy/Cargo.toml:96]: is this unused?

File: shotover-proxy/tests/cassandra_int_tests/functions.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/functions.rs:4]: I know rustfmt isnt running on these files but can we keep manually formatting something like what rustfmt would do?
Its a lot easier to read with the function args split each on their own line instead of the entire function call on one line.

**Comment by conorbros** [comment]
This PR enables us to run `cdrs-tokio` for our Cassandra integration tests. The cpp driver is hidden behind a feature that will also run if the feature is enabled. `rstest` is used to define what tests we want to run each driver for.

There is still some problems running cdrs-tokio on all the tests:
- I will need to get a fix upstreamed to cdrs-tokio to re-enable the collections tests (https://github.com/krojew/cdrs-tokio/pull/103).
- Many of the tests don't work with cdrs-tokio, still need to investigate the reasons for all of these. They have been marked with TODOs

I need to re-enable the collections test before this can be merged.

After that, need to decide whether to:
 - Only merge once cdrs-tokio runs on all the tests. 
-  Merge as is and follow up on problems in later PRs. 

I'm leaning towards merging once cdrs-tokio passes on all the tests.

Getting all the setup code for adding new drivers in could be useful though, for example the scylla driver could also be added.

File: shotover-proxy/tests/cassandra_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/mod.rs:28]: Just sticking a comment here to remind myself this is unaddressed

File: shotover-proxy/Cargo.toml
Comment (rukai) [reply shotover-proxy/Cargo.toml:96]: Oh sorry, searching `rstest` didnt turn up any results, I must have failed to expand the hidden file it was in.
:+1:

**Comment by rukai** [comment]
> I'm leaning towards merging once cdrs-tokio passes on all the tests.

As long as we have integration test cases running on at least the cassandra_cpp driver and we are happy with the multi-driver abstraction then I am happy to merge.

File: shotover-proxy/Cargo.toml
Comment (conorbros) [reply shotover-proxy/Cargo.toml]: Went with `cassandra-cpp-driver-tests` until we need to be generic.

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:303]: Yes. Now that we are testing the error messages generically I added this to make that possible. Plus it makes sense to include this message so might as well.

**Comment by rukai** [comment]
The test run time has halved from 40m -> 20m, I think a bunch of tests are not longer running.

**Comment by conorbros** [comment]
Fixed the issue with it missing some tests, review can continue now.

File: shotover-proxy/tests/cassandra_int_tests/functions.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/functions.rs:34]: I'm not sure how this happened but the changes here are just messing up the formatting.

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (conorbros) [reply shotover-proxy/tests/helpers/cassandra.rs:111]: Just thought it was a bit more convenient since `cassandra-cpp` is sync.

File: shotover-proxy/tests/cassandra_int_tests/collections.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/collections.rs:242]: If we stick a single one of these at the top:
```rust
            let new_set = match driver {
                CassandraDriver::CdrsTokio => ResultValue::Set,
                CassandraDriver::Datastax => ResultValue::List,
            };
```

Then we can implement each of these with something like:
```rust
            assert_query_result(
                session,
                format!(
                    "SELECT my_set FROM test_collections_keyspace.test_set_table_{}_0;",
                    i
                )
                .as_str(),
                &[&[ResultValue::List(vec![new_set(vec![
                    get_type_example_result_value(*native_col_type),
                ])])]],
            )
            .await;
```

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/cassandra.rs:469]: why is this feature flagged out?

Comment (conorbros) [reply shotover-proxy/tests/helpers/cassandra.rs:469]: I definitely had some reason 😂, but not needed anymore. I'll remove that.

Comment (conorbros) [reply shotover-proxy/tests/helpers/cassandra.rs:469]: Because only the cassandra-cpp driver uses it. I'm inclined to remove it but another driver we add might use it.

Comment (rukai) [reply shotover-proxy/tests/helpers/cassandra.rs:469]: I took a look at the protocol spec.
It is a bit of a mess but what I'm reading from it is that varchar and text share the same id: `0x000D` so theres no way to differentiate them at the protocol level.
So lets completely remove the text variant and just map anything that comes up as text into Varchar. 

IMO text is a nicer name and we should just use that instead of varchar everywhere but thats for a future cleanup PR.

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/cassandra.rs:665]: TODO's left

Comment (conorbros) [reply shotover-proxy/tests/helpers/cassandra.rs:665]: Those were already there in the cpp impl. Could we not block this PR on those?

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/cassandra.rs:604]: Can we include in the unimplemented!() a message explaining that text is represented by the same id as varchar at the protocol level and therefore will never be instantiated by the datastax cpp driver

File: shotover-proxy/tests/cassandra_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/mod.rs:565]: I dont want to regress on important testing here.
The error code is the actual important bit here that will alter how the client processes the error so we need to at least be testing that.
Maybe execute_batch_expect_err should return a:
```rust
struct CassandraError {
    code: i32,
    message: String 
}
```
That we map cdrs-tokio and casandra_cpp types into.

The `code` field could be given a custom enum instead of an i32 but not necessary.

File: shotover-proxy/tests/cassandra_int_tests/batch_statements.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/batch_statements.rs:14]: I dont understand why we are changing execute_batch to perform binding?
Arent we better off just using plain old queries?

Comment (conorbros) [reply shotover-proxy/tests/cassandra_int_tests/batch_statements.rs:14]: Hmm I had assumed you couldn't use it that way with `cdrs-tokio` but I might be wrong. Will have a look into that.

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (conorbros) [reply shotover-proxy/tests/helpers/cassandra.rs]: Ended up just switching to async.

File: shotover-proxy/tests/cassandra_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/mod.rs:522]: wouldnt it make sense to just use the enum variant name considering we went to the trouble of setting it up?

File: shotover-proxy/tests/cassandra_int_tests/mod.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/mod.rs:618]: wouldnt it make sense to just use the enum variant name considering we went to the trouble of setting it up?

File: shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs:28]: Our tests here are becoming worse with this change, we no longer test binding a string variable.
But the original tests are pretty lackluster in the first place so im not super worried.

That said, what is your plan for extending execute_prepared into an API that will allow more coverage? Might be worth leaving a TODO so we have some idea of the intention down the line.

If we decide its not worth abstracting the APIs here maybe we could just match on the connection and handle each type seperately?
Otherwise something like this might be the best way forward?
```rust
    pub fn execute_prepared(
        &self,
        prepared_query: &PreparedQuery,
        value: Vec<ResultValue>
    ) -> Vec<Vec<ResultValue>> {
```

File: shotover-proxy/tests/cassandra_int_tests/protect.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/protect.rs:20]: double indent

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/cassandra.rs:212]: maybe just make this branch into a todo!() instead of a bunch of code that doesnt work.

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/cassandra.rs:246]: why dont we first get a mutable reference to schema_awaiter so that we dont have to duplicate this logic?

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/cassandra.rs:274]: deduplicate this logic

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (rukai) [reply shotover-proxy/tests/helpers/cassandra.rs:703]: null and tuple now have implementations for cpp and are very easy to implement here.
I very much agree with Udt though, dont implement that here.

File: shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs
Comment (conorbros) [reply shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs:28]: I could have it similar to how I had the batch queries previously. 

You could pass the query and the variables you wanted to bind and handle for each driver type separately. Would be annoying to handle separate types so we could just test binding ints, which should be enough testing for our purposes.

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (conorbros) [reply shotover-proxy/tests/helpers/cassandra.rs:703]: Sure, I'll implement null and tuple then.

File: shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs
Comment (rukai) [reply shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs:28]: I dont think just testing ints is enough, I think we would want to properly exercise the protocol level handling of different types here.

Gonna go ahead and resolve this, as theres nothing to really block on here, its just future work.

File: shotover-proxy/tests/helpers/cassandra.rs
Comment (conorbros) [reply shotover-proxy/tests/helpers/cassandra.rs:669]: Where's the impl of tuple and null for cpp?

Comment (rukai) [reply shotover-proxy/tests/helpers/cassandra.rs:669]: image

Comment (rukai) [reply shotover-proxy/tests/helpers/cassandra.rs:669]: wow I cant believe rust didnt warn us about the duplicated ValueType::TUPLE :no_mouth: 
image

Comment (conorbros) [reply shotover-proxy/tests/helpers/cassandra.rs:669]: Oh lol. Somehow both were hidden

**Review by rukai — APPROVED** [review]
I think we might end up removing the rstest approach so that we can reuse Cassandra clusters between tests that have identical Cassandra setups.

The rstest approach is fine for now though, let's land this and iterate further in main.

**Review by benbromhead — APPROVED** [review]

**Pull request merged**
